# ANOVA  - Lab

## Introduction

In this lab, you'll get some brief practice generating an ANOVA table (AOV) and interpreting its output. You'll also perform some investigations to compare the method to the t-tests you previously employed to conduct hypothesis testing.

## Objectives

In this lab you will: 

- Use ANOVA for testing multiple pairwise comparisons 
- Interpret results of an ANOVA and compare them to a t-test

## Load the data

Start by loading in the data stored in the file `'ToothGrowth.csv'`: 

In [1]:
import pandas as pd
import statsmodels.api as sm
from statsmodels.formula.api import ols

In [2]:
# Your code here
df = pd.read_csv('ToothGrowth.csv')
df.head()

len supp  dose
0   4.2   VC   0.5
1  11.5   VC   0.5
2   7.3   VC   0.5
3   5.8   VC   0.5
4   6.4   VC   0.5

In [3]:
df

len supp  dose
0    4.2   VC   0.5
1   11.5   VC   0.5
2    7.3   VC   0.5
3    5.8   VC   0.5
4    6.4   VC   0.5
5   10.0   VC   0.5
6   11.2   VC   0.5
7   11.2   VC   0.5
8    5.2   VC   0.5
9    7.0   VC   0.5
10  16.5   VC   1.0
11  16.5   VC   1.0
12  15.2   VC   1.0
13  17.3   VC   1.0
14  22.5   VC   1.0
15  17.3   VC   1.0
16  13.6   VC   1.0
17  14.5   VC   1.0
18  18.8   VC   1.0
19  15.5   VC   1.0
20  23.6   VC   2.0
21  18.5   VC   2.0
22  33.9   VC   2.0
23  25.5   VC   2.0
24  26.4   VC   2.0
25  32.5   VC   2.0
26  26.7   VC   2.0
27  21.5   VC   2.0
28  23.3   VC   2.0
29  29.5   VC   2.0
30  15.2   OJ   0.5
31  21.5   OJ   0.5
32  17.6   OJ   0.5
33   9.7   OJ   0.5
34  14.5   OJ   0.5
35  10.0   OJ   0.5
36   8.2   OJ   0.5
37   9.4   OJ   0.5
38  16.5   OJ   0.5
39   9.7   OJ   0.5
40  19.7   OJ   1.0
41  23.3   OJ   1.0
42  23.6   OJ   1.0
43  26.4   OJ   1.0
44  20.0   OJ   1.0
45  25.2   OJ   1.0
46  25.8   OJ   1.0
47  21.2   OJ   1.0
48  14.5   OJ   1.0
49  27.3   OJ   1.0
50  25.5   OJ   2.0
51  26.4   OJ   2.0
52  22.4   OJ   2.0
53  24.5   OJ   2.0
54  24.8   OJ   2.0
55  30.9   OJ   2.0
56  26.4   OJ   2.0
57  27.3   OJ   2.0
58  29.4   OJ   2.0
59  23.0   OJ   2.0

In [4]:
df.describe()

len       dose
count  60.000000  60.000000
mean   18.813333   1.166667
std     7.649315   0.628872
min     4.200000   0.500000
25%    13.075000   0.500000
50%    19.250000   1.000000
75%    25.275000   2.000000
max    33.900000   2.000000

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   len     60 non-null     float64
 1   supp    60 non-null     object 
 2   dose    60 non-null     float64
dtypes: float64(2), object(1)
memory usage: 1.5+ KB


## Generate the ANOVA table

Now generate an ANOVA table in order to analyze the influence of the medication and dosage:  

In [6]:
# Your code here
formula = 'len ~ C(supp) + dose'
lm = ols(formula, df).fit()
anova_table = sm.stats.anova_lm(lm, typ=2)
print(anova_table)

               sum_sq    df           F        PR(>F)
C(supp)    205.350000   1.0   11.446768  1.300662e-03
dose      2224.304298   1.0  123.988774  6.313519e-16
Residual  1022.555036  57.0         NaN           NaN


## Interpret the output

Make a brief comment regarding the statistics and the effect of supplement and dosage on tooth length: 

# Interpretation of ANOVA Results

## Summary
The ANOVA test reveals that both the supplement type (`supp`) and the dose amount (`dose`) have a **statistically significant effect** on tooth length (`len`).

## Key Results

| Factor | Sum Sq | df | F-value | p-value | Significance |
| :--- | :--- | :--- | :--- | :--- | :--- |
| **C(supp)** | 205.35 | 1.0 | 11.45 | `0.0013` | **Highly Significant** |
| **dose** | 2224.30 | 1.0 | 123.99 | `< 0.0001` | **Extremely Significant** |
| Residual | 1022.56 | 57.0 | NaN | NaN | |

## Detailed Interpretation

### 1. Effect of Supplement Type
- **p-value = 0.0013**
- This is **less than the standard alpha level of 0.05**.
- **Conclusion:** We reject the null hypothesis. There is a statistically significant difference in tooth length between the Orange Juice (OJ) and Ascorbic Acid (VC) supplement groups.

### 2. Effect of Dose
- **p-value < 0.0001**
- This is **highly significant**, far below the 0.05 threshold.
- **Conclusion:** We reject the null hypothesis. The dosage level has a very strong and statistically significant effect on tooth growth.

### 3. Effect Size
- The **Sum of Squares (Sum Sq)** for `dose` (2224.30) is much larger than for `supp` (205.35).
- This indicates that **dose amount explains more variance** in tooth length than supplement type does.

## Final Conclusion
Both factors are important predictors of tooth growth. The dosage amount is the dominant factor, but the type of supplement also plays a significant role, even after accounting for the dose.

## Compare to t-tests

Now that you've had a chance to generate an ANOVA table, its interesting to compare the results to those from the t-tests you were working with earlier. With that, start by breaking the data into two samples: those given the OJ supplement, and those given the VC supplement. Afterward, you'll conduct a t-test to compare the tooth length of these two different samples: 

In [7]:
# Your code here
# Separate the data into OJ and VC groups
oj_data = df[df['supp'] == 'OJ']['len']
vc_data = df[df['supp'] == 'VC']['len']

print(f"OJ group: n = {len(oj_data)}, mean = {oj_data.mean():.2f}")
print(f"VC group: n = {len(vc_data)}, mean = {vc_data.mean():.2f}")

OJ group: n = 30, mean = 20.66
VC group: n = 30, mean = 16.96


Now run a t-test between these two groups and print the associated two-sided p-value: 

In [8]:
# Calculate the 2-sided p-value for a t-test comparing the two supplement groups
from scipy import stats

# Perform independent t-test
t_stat, p_value = stats.ttest_ind(oj_data, vc_data)

print(f"T-test results:")
print(f"T-statistic: {t_stat:.4f}")
print(f"P-value: {p_value:.6f}")

T-test results:
T-statistic: 1.9153
P-value: 0.060393


## A 2-Category ANOVA F-test is equivalent to a 2-tailed t-test!

Now, recalculate an ANOVA F-test with only the supplement variable. An ANOVA F-test between two categories is the same as performing a 2-tailed t-test! So, the p-value in the table should be identical to your calculation above.

> Note: there may be a small fractional difference (>0.001) between the two values due to a rounding error between implementations. 

In [9]:
# Your code here; conduct an ANOVA F-test of the oj and vc supplement groups.
# Perform one-way ANOVA for supplement only
model_supp_only = ols('len ~ C(supp)', data=df).fit()
anova_table_supp_only = sm.stats.anova_lm(model_supp_only, typ=2)

print("One-way ANOVA results for supplement only:")
print(anova_table_supp_only)
# Compare the p-value to that of the t-test above. 
# They should match (there may be a tiny fractional difference due to rounding errors in varying implementations)

One-way ANOVA results for supplement only:
               sum_sq    df         F    PR(>F)
C(supp)    205.350000   1.0  3.668253  0.060393
Residual  3246.859333  58.0       NaN       NaN


In [10]:
# Compare the p-value to that of the t-test above. 
# Print both results for clear comparison
print("COMPARISON: One-Way ANOVA (supp only) vs. Independent T-Test")
print("="*60)
print(f"{'ONE-WAY ANOVA (supp only)':<30} {'INDEPENDENT T-TEST':<30}")
print(f"{'F-value:':<20} {3.668253:<10.4f} {'T-statistic:':<20} {1.9153:<10.4f}")
print(f"{'P-value:':<20} {0.060393:<10.6f} {'P-value:':<20} {0.060393:<10.6f}")
print("="*60)
print("✓ P-values are IDENTICAL (0.060393)")
print("✓ Mathematical relationship holds: F = t² (3.668 ≈ 1.9153²)")

COMPARISON: One-Way ANOVA (supp only) vs. Independent T-Test
ONE-WAY ANOVA (supp only)      INDEPENDENT T-TEST            
F-value:             3.6683     T-statistic:         1.9153    
P-value:             0.060393   P-value:             0.060393  
✓ P-values are IDENTICAL (0.060393)
✓ Mathematical relationship holds: F = t² (3.668 ≈ 1.9153²)


## Run multiple t-tests

While the 2-category ANOVA test is identical to a 2-tailed t-test, performing multiple t-tests leads to the multiple comparisons problem. To investigate this, look at the various sample groups you could create from the 2 features: 

In [11]:
for group in df.groupby(['supp', 'dose'])['len']:
    group_name = group[0]
    data = group[1]
    print(group_name)

('OJ', np.float64(0.5))
('OJ', np.float64(1.0))
('OJ', np.float64(2.0))
('VC', np.float64(0.5))
('VC', np.float64(1.0))
('VC', np.float64(2.0))


While bad practice, examine the effects of calculating multiple t-tests with the various combinations of these. To do this, generate all combinations of the above groups. For each pairwise combination, calculate the p-value of a 2-sided t-test. Print the group combinations and their associated p-value for the two-sided t-test.

In [12]:
# Your code here; reuse your t-test code above to calculate the p-value for a 2-sided t-test
# for all combinations of the supplement-dose groups listed above. 
# (Since there isn't a control group, compare each group to every other group.)
from itertools import combinations

# Get all unique group combinations
groups = list(df.groupby(['supp', 'dose'])['len'])
group_names = [group[0] for group in groups]
group_data = [group[1] for group in groups]

print("All possible group combinations:")
for i, name in enumerate(group_names):
    print(f"{i}: {name}")

print("\n" + "="*60)
print("PAIRWISE T-TEST RESULTS (TWO-SIDED)")
print("="*60)

# Generate all pairwise combinations
for (i, j) in combinations(range(len(group_names)), 2):
    group1_name = group_names[i]
    group2_name = group_names[j]
    data1 = group_data[i]
    data2 = group_data[j]
    
    # Perform t-test
    t_stat, p_value = stats.ttest_ind(data1, data2)
    
    print(f"{group1_name} vs {group2_name}:")
    print(f"  p-value = {p_value:.6f}")
    print(f"  Significant at α=0.05? {'YES' if p_value < 0.05 else 'NO'}")
    print("-" * 40)

All possible group combinations:
0: ('OJ', np.float64(0.5))
1: ('OJ', np.float64(1.0))
2: ('OJ', np.float64(2.0))
3: ('VC', np.float64(0.5))
4: ('VC', np.float64(1.0))
5: ('VC', np.float64(2.0))

PAIRWISE T-TEST RESULTS (TWO-SIDED)
('OJ', np.float64(0.5)) vs ('OJ', np.float64(1.0)):
  p-value = 0.000084
  Significant at α=0.05? YES
----------------------------------------
('OJ', np.float64(0.5)) vs ('OJ', np.float64(2.0)):
  p-value = 0.000000
  Significant at α=0.05? YES
----------------------------------------
('OJ', np.float64(0.5)) vs ('VC', np.float64(0.5)):
  p-value = 0.005304
  Significant at α=0.05? YES
----------------------------------------
('OJ', np.float64(0.5)) vs ('VC', np.float64(1.0)):
  p-value = 0.042240
  Significant at α=0.05? YES
----------------------------------------
('OJ', np.float64(0.5)) vs ('VC', np.float64(2.0)):
  p-value = 0.000007
  Significant at α=0.05? YES
----------------------------------------
('OJ', np.float64(1.0)) vs ('OJ', np.float64(2.0)):
 

## Summary

In this lesson, you implemented the ANOVA technique to generalize testing methods to multiple groups and factors.